File: contrib/flir-seq/test/compare-flirpy.mjs
// Copyright 2026 The Gitea Authors. All rights reserved.
// SPDX-License-Identifier: MIT
//
// Cross-checks the viewer's radiometry against flirpy, an independent Python
// implementation of the same FLIR model, over a grid of parameter sets.
//
//   pip install flirpy
//   node compare-flirpy.mjs                    parameter grid
//   node compare-flirpy.mjs real.seq           also every pixel of a real file
//   FLIRPY_PYTHON=/path/to/venv/bin/python node compare-flirpy.mjs
//
// flirpy has no equivalent of estAtmosphericTransmission, so every case here
// leaves it at 0 (the transmission is estimated). That branch is covered by
// run.mjs instead.

import {execFileSync} from 'node:child_process';
import {mkdtempSync, writeFileSync, readFileSync, rmSync} from 'node:fs';
import {tmpdir} from 'node:os';
import {join, dirname} from 'node:path';
import {fileURLToPath} from 'node:url';
import {loadFlirSeq} from './load.mjs';
import {toArrayBuffer} from './fixture.mjs';

const here = dirname(fileURLToPath(import.meta.url));
const python = process.env.FLIRPY_PYTHON || 'python3';
const flir = loadFlirSeq();
const TOLERANCE = 1e-9; // Kelvin

// The A655sc calibration the samples carry, as the baseline every case varies.
const base = flir.paramsFromInfo({
  emissivity: 0.95, objectDistance: 1, reflectedTempK: 293.15, atmosphericTempK: 293.15,
  irWindowTempK: 293.15, irWindowTransmission: 1, relativeHumidity: 50,
  planckR1: 14772.65, planckR2: 0.0137111, planckB: 1393.8, planckF: 1, planckO: -3735,
  atmTransAlpha1: 0.006569, atmTransAlpha2: 0.01262,
  atmTransBeta1: -0.002276, atmTransBeta2: -0.00667, atmTransX: 1.9,
});

const variations = [
  ['A655sc, as recorded', {}],
  ['low emissivity', {emissivity: 0.3}],
  ['high emissivity', {emissivity: 1}],
  ['hot reflected background', {reflectedTemp: 80}],
  ['cold reflected background', {reflectedTemp: -40}],
  ['long distance', {objectDistance: 250}],
  ['very long distance', {objectDistance: 1000}],
  ['dry air', {relativeHumidity: 5}],
  ['saturated air', {relativeHumidity: 100}],
  ['hot air', {objectDistance: 100, atmosphericTemp: 45}],
  ['freezing air', {objectDistance: 100, atmosphericTemp: -30}],
  ['IR window in the path', {irWindowTransmission: 0.7, irWindowTemp: 35}],
  ['everything at once', {
    emissivity: 0.62, objectDistance: 42, relativeHumidity: 88,
    reflectedTemp: 12, atmosphericTemp: 31, irWindowTemp: 26, irWindowTransmission: 0.85,
  }],
  ['a different camera calibration', {
    planckR1: 17096.453, planckR2: 0.04351538, planckB: 1428.0, planckF: 1, planckO: -7340,
  }],
];

// raw counts spanning what the samples hold, plus the extremes of the domain
const raws = [4000, 7459, 8500, 9891, 10500, 11674, 14000, 20000, 40000];

const cases = variations.map(([name, override]) => ({
  name,
  params: Object.assign({}, base, override),
  raw: raws,
}));

const work = mkdtempSync(join(tmpdir(), 'flir-seq-flirpy-'));
let theirs;
try {
  const casePath = join(work, 'cases.json');
  const outPath = join(work, 'out.json');
  writeFileSync(casePath, JSON.stringify(cases));
  try {
    execFileSync(python, [join(here, 'compare_flirpy.py'), casePath, outPath], {stdio: ['ignore', 'inherit', 'inherit']});
  } catch (e) {
    console.error('could not run flirpy through "' + python + '": ' + e.message);
    console.error('install it with:  pip install flirpy   (or set FLIRPY_PYTHON)');
    process.exit(2);
  }
  theirs = JSON.parse(readFileSync(outPath, 'utf8'));
} finally {
  rmSync(work, {recursive: true, force: true});
}

let worst = 0;
let worstCase = '';
const fmt = (v) => (Number.isFinite(v) ? v.toFixed(6) : 'out of domain').padStart(13);
console.log('case                          raw     flir-seq        flirpy          |Δ|');
for (let i = 0; i < cases.length; i++) {
  const conv = flir.makeConverter(cases[i].params);
  if (!conv.ok) throw new Error('case ' + cases[i].name + ' has no usable calibration: ' + conv.reason);
  let caseWorst = 0, at = raws[0], mine = conv.toTemp(raws[0]);
  let other = theirs[i][0] === null ? NaN : theirs[i][0];
  for (let j = 0; j < raws.length; j++) {
    const a = conv.toTemp(raws[j]);
    // null is flirpy's NaN: a raw value the model cannot invert. Both sides
    // refusing is agreement; one side inventing a number is not.
    const b = theirs[i][j] === null ? NaN : theirs[i][j];
    const bothOutOfDomain = !Number.isFinite(a) && !Number.isFinite(b);
    const d = bothOutOfDomain ? 0 : (Number.isFinite(a) && Number.isFinite(b) ? Math.abs(a - b) : Infinity);
    if (d > caseWorst) {
      caseWorst = d;
      at = raws[j];
      mine = a;
      other = b;
    }
  }
  console.log(cases[i].name.padEnd(29) + String(at).padStart(5) + '  ' +
    fmt(mine) + '  ' + fmt(other) + '  ' + caseWorst.toExponential(2));
  if (caseWorst > worst) {
    worst = caseWorst;
    worstCase = cases[i].name;
  }
}

console.log('\nflirpy ' + execFileSync(python,
  ['-c', 'import importlib.metadata as m; print(m.version("flirpy"))'], {encoding: 'utf8'}).trim() +
  ', ' + cases.length + ' parameter sets × ' + raws.length + ' raw values');
let ok = worst <= TOLERANCE;
console.log((ok ? 'agrees' : 'DISAGREES') + ': largest deviation ' + worst.toExponential(2) +
  ' K (' + worstCase + '), tolerance ' + TOLERANCE.toExponential(0) + ' K');

// --- whole files, pixel by pixel ------------------------------------------
//
// This puts flirpy's own FFF reader against ours, so the record offsets are
// compared as well as the arithmetic.
for (const path of process.argv.slice(2)) {
  console.log('\n' + path);
  const dir = mkdtempSync(join(tmpdir(), 'flir-seq-flirpy-'));
  let decoded;
  try {
    const outPath = join(dir, 'frames.json');
    execFileSync(python, [join(here, 'compare_flirpy.py'), '--frames', path, outPath],
      {stdio: ['ignore', 'inherit', 'inherit']});
    decoded = JSON.parse(readFileSync(outPath, 'utf8'));
  } finally {
    rmSync(dir, {recursive: true, force: true});
  }

  const buffer = toArrayBuffer(readFileSync(path));
  const parsed = flir.parseSeq(buffer);
  const agree = (label, condition, detail) => {
    console.log((condition ? '  ok   ' : '  FAIL ') + label + (detail === undefined ? '' : '  ' + detail));
    if (!condition) ok = false;
  };
  agree('same frame count', parsed.frames.length === decoded.length,
    parsed.frames.length + ' vs ' + decoded.length);

  for (let i = 0; i < Math.min(parsed.frames.length, decoded.length); i++) {
    const frame = parsed.frames[i];
    const theirFrame = decoded[i];
    // flirpy's FFF reader converts Kelvin to Celsius with 273.14 (fff.py's
    // get_float_kelvin) while everything else in it uses 273.15, so its object
    // parameters land 0.01 K high. Comparing twice separates that from a real
    // disagreement: once with flirpy's own numbers fed to our converter, once
    // with each side's own parsing.
    const ours = flir.paramsFromInfo(frame.info);
    const sameInputs = Object.assign({}, ours, {
      reflectedTemp: theirFrame.meta['Reflected Apparent Temperature'],
      atmosphericTemp: theirFrame.meta['Atmospheric Temperature'],
      irWindowTemp: theirFrame.meta['IR Window Temperature'],
    });
    const convOurs = flir.makeConverter(ours);
    const convSame = flir.makeConverter(sameInputs);
    const pixels = flir.readFramePixels(buffer, frame);
    let frameWorst = 0, asParsed = 0, mismatched = 0;
    for (let j = 0; j < pixels.length; j++) {
      const a = convSame.toTemp(pixels[j]);
      const b = theirFrame.temps[j] === null ? NaN : theirFrame.temps[j];
      if (!Number.isFinite(a) || !Number.isFinite(b)) {
        if (Number.isFinite(a) !== Number.isFinite(b)) mismatched++;
        continue;
      }
      const d = Math.abs(a - b);
      if (d > frameWorst) frameWorst = d;
      const dp = Math.abs(convOurs.toTemp(pixels[j]) - b);
      if (dp > asParsed) asParsed = dp;
    }
    agree('frame ' + (i + 1) + ': same offset and geometry',
      frame.offset === theirFrame.offset && frame.raw.width === theirFrame.width &&
      frame.raw.height === theirFrame.height,
      frame.offset + ' ' + frame.raw.width + 'x' + frame.raw.height);
    agree('frame ' + (i + 1) + ': ' + pixels.length + ' pixels agree',
      frameWorst <= TOLERANCE && mismatched === 0,
      'largest deviation ' + frameWorst.toExponential(2) + ' K; ' +
      asParsed.toExponential(2) + " K with each side's own Kelvin constant" +
      (mismatched ? ', ' + mismatched + ' disagree on being out of domain' : ''));
    if (frameWorst > worst) worst = frameWorst;
  }
}

process.exit(ok ? 0 : 1);

File: contrib/flir-seq/test/compare_flirpy.py
# Copyright 2026 The Gitea Authors. All rights reserved.
# SPDX-License-Identifier: MIT
#
# Helper for compare-flirpy.mjs: reads the case file it writes, runs each case
# through flirpy's raw2temp() and writes the temperatures back as JSON.
#
#   python3 compare_flirpy.py cases.json out.json
#   python3 compare_flirpy.py --frames real.seq out.json
#
# flirpy takes its parameters the way its own FFF reader produces them:
# temperatures in Celsius and the relative humidity as a fraction, which is
# what the camera stores at 0x3c.

import json
import math
import sys

import numpy as np
from flirpy.util.raw import raw2temp


def main(case_path, out_path):
    cases = json.load(open(case_path))
    results = []
    for case in cases:
        # flirpy's from_string_or_float() accepts a str or a float but not an
        # int, so every number has to arrive as a float.
        p = {k: (float(v) if isinstance(v, (int, float)) else v) for k, v in case["params"].items()}
        meta = {
            "Emissivity": p["emissivity"],
            "Object Distance": p["objectDistance"],
            "Reflected Apparent Temperature": p["reflectedTemp"],
            "Atmospheric Temperature": p["atmosphericTemp"],
            "IR Window Temperature": p["irWindowTemp"],
            "IR Window Transmission": p["irWindowTransmission"],
            # flirpy's own FFF reader hands this on as the fraction from 0x3c
            "Relative Humidity": p["relativeHumidity"] / 100.0,
            "Planck R1": p["planckR1"],
            "Planck R2": p["planckR2"],
            "Planck B": p["planckB"],
            "Planck F": p["planckF"],
            "Planck O": p["planckO"],
            "Atmospheric Trans Alpha 1": p["atmTransAlpha1"],
            "Atmospheric Trans Alpha 2": p["atmTransAlpha2"],
            "Atmospheric Trans Beta 1": p["atmTransBeta1"],
            "Atmospheric Trans Beta 2": p["atmTransBeta2"],
            "Atmospheric Trans X": p["atmTransX"],
        }
        temps = raw2temp(np.array(case["raw"], dtype=np.float64), meta)
        # NaN is not valid JSON; a raw value outside the model's domain comes
        # back as null and has to be NaN on the other side too.
        results.append([float(t) if math.isfinite(t) else None for t in temps])
    json.dump(results, open(out_path, "w"))


def frames(seq_path, out_path):
    """Decode a real .seq end to end with flirpy's own FFF reader.

    A .seq is a concatenation of FFF frames, and flirpy only reads one, so the
    frames are split on the magic first and handed over one at a time. This
    exercises flirpy's parser as well as its radiometry, which is the point:
    the comparison then covers the offsets too, not just the formula.
    """
    from flirpy.io.fff import Fff

    data = open(seq_path, "rb").read()
    magic = b"FFF\x00"
    starts = []
    at = data.find(magic)
    while at != -1:
        starts.append(at)
        at = data.find(magic, at + 1)
    starts.append(len(data))

    out = []
    for i in range(len(starts) - 1):
        fff = Fff(data[starts[i]:starts[i + 1]])
        image = fff.get_radiometric_image()
        out.append({
            "offset": starts[i],
            "width": int(fff.width),
            "height": int(fff.height),
            # flirpy's own parse of the sensor counts, to compare the record
            # offsets independently of any radiometry
            "raw": [int(v) for v in fff.get_image().reshape(-1).tolist()],
            "meta": {k: (v if isinstance(v, (int, float)) else str(v)) for k, v in fff.meta.items()
                     if isinstance(v, (int, float, bytes, str))},
            "temps": [float(t) if math.isfinite(t) else None for t in image.reshape(-1).tolist()],
        })
    json.dump(out, open(out_path, "w"))


if __name__ == "__main__":
    if sys.argv[1] == "--frames":
        frames(sys.argv[2], sys.argv[3])
    else:
        main(sys.argv[1], sys.argv[2])
